src/engine/include/includejs/engine_private_data.h

```diff
@@ -0,0 +1,46 @@
+
+#ifndef INCLUDEJS_ENGINE_PRIVATE_DATA_H
+#define INCLUDEJS_ENGINE_PRIVATE_DATA_H
+
+#include <includejs/engine_value.h>
+
+#include <map>    // std::map
+#include <memory> // std::unique_ptr
+
+namespace includejs {
+
+// Inspired by https://github.com/sourcemeta/jsontoolkit
+/// @ingroup engine
+struct INCLUDEJS_ENGINE_EXPORT PrivateObjectData {
+  std::unique_ptr<std::map<void *, Value::Function>> functions;
+
+  PrivateObjectData() {
+    functions = std::make_unique<std::map<void *, Value::Function>>();
+  }
+
+  auto data() -> void * { return data_; }
+
+  auto set_data(void *new_data, std::function<void(void *)> new_deleter)
```

Comment:
I left some comments on the Starship PR. I guess the same ones apply here? Not sure which PR you want to do first

Reply:
lets fix the other one first.